nav2_util/test/test_actions.cpp

```diff
@@ -0,0 +1,331 @@
+// Copyright (c) 2018 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <chrono>
+#include <memory>
+#include <thread>
+
+#include "gtest/gtest.h"
+#include "nav2_util/node_utils.hpp"
+#include "nav2_util/simple_action_client.hpp"
+#include "nav2_util/simple_action_server.hpp"
+#include "test_msgs/action/fibonacci.hpp"
+
+using Fibonacci = test_msgs::action::Fibonacci;
+using GoalHandle = rclcpp_action::ServerGoalHandle<Fibonacci>;
+
+using std::placeholders::_1;
+using namespace std::chrono_literals;
+
+class FibonacciServerNode : public rclcpp::Node
+{
+public:
+  FibonacciServerNode()
+  : rclcpp::Node("fibonacci_server_node")
+  {
+  }
+
+  void on_init()
+  {
+    action_server_ = std::make_shared<nav2_util::SimpleActionServer<Fibonacci>>(
+      shared_from_this(),
+      "fibonacci",
+      std::bind(&FibonacciServerNode::execute, this, std::placeholders::_1));
+  }
+
+  void on_term()
+  {
+    action_server_.reset();
+  }
+
+  void execute(const std::shared_ptr<GoalHandle> goal_handle)
+  {
+    // The goal may be pre-empted, so keep a pointer to the current goal
+    std::shared_ptr<GoalHandle> current_goal_handle = goal_handle;
+
+preempted:
+    // Initialize the goal, feedback, and result
+    auto goal = current_goal_handle->get_goal();
+    auto feedback = std::make_shared<Fibonacci::Feedback>();
+    auto result = std::make_shared<Fibonacci::Result>();
+
+    // Fibonacci-specific initialization
+    rclcpp::Rate loop_rate(5);
+    auto & sequence = feedback->sequence;
+    sequence.push_back(0);
+    sequence.push_back(1);
+
+    for (int i = 1; (i < goal->order) && rclcpp::ok(); ++i) {
+      // Check if this action has been canceled
+      if (current_goal_handle->is_canceling()) {
+        result->sequence = sequence;
+        current_goal_handle->set_canceled(result);
+        return;
+      }
+
+      // Check if we've gotten an new goal, pre-empting the current one
+      if (action_server_->update_requested()) {
+        current_goal_handle = action_server_->get_updated_goal_handle();
+        goto preempted;
```

Comment:
Good point about the loop rate. That could be hoisted above the preempted: label. I'll make that change.

Regarding the suggested change: It's not just the goal that has to be updated, but in the general case, one must also clear the feedback and potentially the result (it could be possible that a result was being accumulated). So, there would need to be code to handle this as well; basically a duplication of the code that is already present above. So, I think that it's less error prone and a better template the way it is.

For example, this:

```C++
  void execute(const std::shared_ptr<GoalHandle> goal_handle)
  {
    // The goal may be pre-empted, so keep a pointer to the current goal
    std::shared_ptr<GoalHandle> current_goal_handle = goal_handle;

    rclcpp::Rate loop_rate(10);

preempted:
    // Initialize the goal, feedback, and result
    auto goal = current_goal_handle->get_goal();
    auto feedback = std::make_shared<Fibonacci::Feedback>();
    auto result = std::make_shared<Fibonacci::Result>();

    // Fibonacci-specific initialization
    auto & sequence = feedback->sequence;
    sequence.push_back(0);
    sequence.push_back(1);

    for (int i = 1; (i < goal->order) && rclcpp::ok(); ++i) {
      // Check if this action has been canceled
      ...

      // Check if we've gotten an new goal, pre-empting the current one
      if (action_server_->update_requested()) {
        current_goal_handle = action_server_->get_updated_goal_handle();
        goto preempted;
      }

      // Update the sequence
      sequence.push_back(sequence[i] + sequence[i - 1]);

      // Publish feedback
      current_goal_handle->publish_feedback(feedback);
      loop_rate.sleep();
    }
    ...
  }
```

Instead of this:

```C++
  void execute(const std::shared_ptr<GoalHandle> goal_handle)
  {
    // The goal may be pre-empted, so keep a pointer to the current goal
    std::shared_ptr<GoalHandle> current_goal_handle = goal_handle;

    rclcpp::Rate loop_rate(10);

    // Initialize the goal, feedback, and result
    auto goal = current_goal_handle->get_goal();
    auto feedback = std::make_shared<Fibonacci::Feedback>();
    auto result = std::make_shared<Fibonacci::Result>();

    // Fibonacci-specific initialization
    auto & sequence = feedback->sequence;
    sequence.push_back(0);
    sequence.push_back(1);

    for (int i = 1; (i < goal->order) && rclcpp::ok(); ++i) {
      // Check if this action has been canceled
      ...

      // Check if we've gotten an new goal, pre-empting the current one
      if (action_server_->update_requested()) {
        current_goal_handle = action_server_->get_updated_goal_handle();
        goal = current_goal_handle->get_goal();
        feedback = std::make_shared<Fibonacci::Feedback>();
        result = std::make_shared<Fibonacci::Result>(); 

        sequence = feedback->sequence;
        sequence.push_back(0);
        sequence.push_back(1);
      } else {
          // Update the sequence
          sequence.push_back(sequence[i] + sequence[i - 1]);

          // Publish feedback
          current_goal_handle->publish_feedback(feedback);
      }

      loop_rate.sleep();
    }
    ...
  }
```

Reply:
By the way, in some cases, where there isn't any state in the feedback and result, I agree with the change you propose. DWB, for example is like this. So, I think which way is cleaner will depend on the circumstances.